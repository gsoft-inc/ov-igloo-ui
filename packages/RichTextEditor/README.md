# RichTextEditor

  TODO: Write your component description here.

  <ReferenceLinks is="custom" />

  ## Installation

  To install `@igloo-ui/rich-text-editor` in your project, you will need to run the following command using [npm](https://www.npmjs.com/):

  ```bash
  npm install @igloo-ui/rich-text-editor
  ```

  If you prefer [Yarn](https://classic.yarnpkg.com/en/), use the following command instead:

  ```bash
  yarn add @igloo-ui/rich-text-editor
  ```

  ## Usage

  Then to use the component in your code just import it!

  ```jsx
  import RichTextEditor from '@igloo-ui/rich-text-editor';

  // TODO: Add your component usage here
  ```
  