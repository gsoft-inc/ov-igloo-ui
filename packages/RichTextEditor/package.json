{
  "name": "@igloo-ui/rich-text-editor",
  "version": "0.0.0",
  "main": "dist/RichTextEditor.js",
  "module": "dist/RichTextEditor.js",
  "types": "dist/RichTextEditor.d.ts",
  "styles": "dist/rich-text-editor.css",
  "publishConfig": {
    "access": "public"
  },
  "repository": "git+https://github.com/gsoft-inc/ov-igloo-ui.git",
  "author": "Officevibe Inc.",
  "contributors": [
    "Vicky Comeau vicky.comeau@gsoft.com"
  ],
  "license": "Apache-2.0",
  "peerDependencies": {
    "react": ">= 16.8.6"
  },
  "files": [
    "dist"
  ],
  "scripts": {
    "build": "rollup -c rollup.config.js"
  },
  "dependencies": {
    "@igloo-ui/button": "^0.4.1",
    "@igloo-ui/icon-button": "^1.0.1",
    "@igloo-ui/input": "^2.0.4",
    "@igloo-ui/tokens": "^2.0.0",
    "@igloo-ui/tooltip": "^3.1.3",
    "@lexical/code": "^0.10.0",
    "@lexical/link": "^0.10.0",
    "@lexical/list": "^0.10.0",
    "@lexical/markdown": "^0.10.0",
    "@lexical/overflow": "^0.10.0",
    "@lexical/react": "^0.10.0",
    "@lexical/rich-text": "^0.10.0",
    "@lexical/selection": "^0.10.0",
    "@lexical/utils": "^0.10.0",
    "@types/react-dom": "18.0.11",
    "classnames": "^2.3.2",
    "lexical": "^0.10.0"
  },
  "browserslist": [
    "> 1%",
    "last 2 versions",
    "IE >= 11"
  ]
}
