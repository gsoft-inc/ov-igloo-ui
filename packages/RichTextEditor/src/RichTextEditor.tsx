import * as React from 'react';

import cx from 'classnames';

import { EditorState, LexicalEditor } from 'lexical';
import {
  LexicalComposer,
  InitialConfigType,
} from '@lexical/react/LexicalComposer';
import { RichTextPlugin } from '@lexical/react/LexicalRichTextPlugin';
import { ContentEditable } from '@lexical/react/LexicalContentEditable';
import { HistoryPlugin } from '@lexical/react/LexicalHistoryPlugin';
import { AutoFocusPlugin } from '@lexical/react/LexicalAutoFocusPlugin';
import LexicalErrorBoundary from '@lexical/react/LexicalErrorBoundary';
// eslint-disable-next-line max-len
import { CharacterLimitPlugin } from '@lexical/react/LexicalCharacterLimitPlugin';
import { ListPlugin } from '@lexical/react/LexicalListPlugin';
// eslint-disable-next-line max-len
import { MarkdownShortcutPlugin } from '@lexical/react/LexicalMarkdownShortcutPlugin';
import { ClearEditorPlugin } from '@lexical/react/LexicalClearEditorPlugin';
import { OnChangePlugin } from '@lexical/react/LexicalOnChangePlugin';

import { HeadingNode, QuoteNode } from '@lexical/rich-text';
import { ListItemNode, ListNode } from '@lexical/list';
import { AutoLinkNode, LinkNode } from '@lexical/link';
import { OverflowNode } from '@lexical/overflow';
import { CodeNode, CodeHighlightNode } from '@lexical/code';

import { TRANSFORMERS } from '@lexical/markdown';

import Lock from '@igloo-ui/icons/dist/Lock';
import ToolbarPlugin from './plugins/ToolbarPlugin';
import { OnFocusPlugin } from './plugins/OnFocusPlugin';
import { DisablePlugin } from './plugins/DisablePlugin';
import { MaxLengthPlugin } from './plugins/MaxLengthPlugin';
import FloatingLinkEditorPlugin from './plugins/FloatingLinkEditorPlugin';
import LinkPlugin from './plugins/LinkPlugin';
import CodeHighlightPlugin from './plugins/CodeHighlightPlugin';

import EditorTheme from './themes/RichTextEditor.theme';

import './rich-text-editor.scss';

export interface MessageOptions {
  text?: string;
  tooltip?: string;
}

export interface Messages {
  bold?: MessageOptions;
  clear?: MessageOptions;
  italic?: MessageOptions;
  link?: MessageOptions;
  orderedList?: MessageOptions;
  strikethrough?: MessageOptions;
  underline?: MessageOptions;
  unorderedList?: MessageOptions;
  private?: MessageOptions;
  linkEditorEdit?: MessageOptions;
  linkEditorRemove?: MessageOptions;
  linkEditorCancel?: MessageOptions;
  linkEditorSave?: MessageOptions;
}

export interface RichTextEditorProps
  extends Omit<React.ComponentProps<'div'>, 'onChange' | 'onFocus' | 'onBlur'> {
  /** Whether or not the editor should be focused on load */
  autoFocus?: boolean;
  /** Add a class name to the rich text editor */
  className?: string;
  /** Add a data-test tag for automated tests */
  dataTest?: string;
  /** Whether or not the rich text editor is disabled */
  disabled?: boolean;
  /** Whether or not the right text editor has an error */
  error?: boolean;
  /** The initial editor state, which will preload the editor with content */
  initialState?: string;
  /** Whether or not the text can be cleared */
  isClearable?: boolean;
  /** Whether or not the rich text editor should have a private look */
  isPrivate?: boolean;
  /** The maximum amount or characters allowed
   * (Automatically adds a character limit indicator) */
  maxLength?: number;
  /** Messages for the buttons */
  messages?: Messages;
  /** The event that is called when the editor loses focus */
  onBlur?: (editor?: LexicalEditor) => void;
  /** The event that is called when the editor text changes */
  onChange?: (editorState: string) => void;
  /** The event that's called when the editor gains focus */
  onFocus?: (editor?: LexicalEditor) => void;
  /** The placeholder text to display int he editor */
  placeholder?: string;
  /** Usually the save button for the editor */
  primaryBtn?: React.ReactElement;
  /** Wether or not the editor should display the toolbar */
  showToolbar?: boolean;
  /** Wether or not the editor should display
   * the toolbar when editor is focused */
  showToolbarOnFocus?: boolean;
}

const RichTextEditor: React.FunctionComponent<RichTextEditorProps> = ({
  autoFocus = false,
  className,
  dataTest,
  disabled,
  error,
  initialState,
  isClearable = true,
  isPrivate,
  maxLength,
  messages = {
    bold: { tooltip: 'Bold' },
    clear: { text: 'Clear', tooltip: 'Clear all text' },
    italic: { tooltip: 'Italic' },
    link: { tooltip: 'Link' },
    orderedList: { tooltip: 'Ordered List' },
    strikethrough: { tooltip: 'Strikethrough' },
    underline: { tooltip: 'Underline' },
    unorderedList: { tooltip: 'Unordered List' },
    private: { text: 'Visible to you only' },
    linkEditorEdit: { tooltip: 'Edit link' },
    linkEditorCancel: { tooltip: 'Cancel' },
    linkEditorRemove: { tooltip: 'Remove link' },
    linkEditorSave: { tooltip: 'Save' },
  },
  onBlur,
  onChange,
  onFocus,
  placeholder,
  primaryBtn,
  showToolbar = true,
  showToolbarOnFocus = false,
}: RichTextEditorProps) => {
  const [hasFocus, setHasFocus] = React.useState(autoFocus);

  const [floatingAnchorElem, setFloatingAnchorElem] =
    React.useState<HTMLDivElement | null>(null);

  const onRef = (_floatingAnchorElem: HTMLDivElement): void => {
    if (_floatingAnchorElem !== null) {
      setFloatingAnchorElem(_floatingAnchorElem);
    }
  };

  function Placeholder(): JSX.Element {
    return (
      <div className="ids-rich-text-editor__placeholder">{placeholder}</div>
    );
  }

  const editorConfig: InitialConfigType = {
    namespace: 'ids-rich-text-editor',
    theme: EditorTheme,
    editorState: initialState,
    editable: !disabled,
    // Handling of errors during update
    onError(error) {
      throw error;
    },
    // Any custom nodes go here
    nodes: [
      ListNode,
      ListItemNode,
      AutoLinkNode,
      LinkNode,
      HeadingNode,
      QuoteNode,
      CodeNode,
      CodeHighlightNode,
      OverflowNode,
    ],
  };

  const classes = cx('ids-rich-text-editor', className, {
    'ids-rich-text-editor--error': error,
    'ids-rich-text-editor--private': isPrivate,
    'ids-rich-text-editor--focus': hasFocus,
    'ids-rich-text-editor--disabled': disabled,
  });

  const handleOnChange = (editorState: EditorState): void => {
    onChange?.(JSON.stringify(editorState));
  };

  const handleOnFocus = (editor: LexicalEditor): void => {
    setHasFocus(true);
    onFocus?.(editor);
  };

  const handleOnBlur = (editor: LexicalEditor): void => {
    setHasFocus(false);
    onBlur?.(editor);
  };

  const showFooter = !!maxLength || isPrivate || !!primaryBtn;

  return (
    <LexicalComposer initialConfig={editorConfig}>
      <div className={classes} data-test={dataTest}>
        {showToolbar && (
          <ToolbarPlugin
            disabled={disabled}
            isClearable={isClearable}
            messages={messages}
            showOnFocus={showToolbarOnFocus}
          />
        )}
        <div className="ids-rich-text-editor__inner">
          <RichTextPlugin
            contentEditable={
              <div className="ids-rich-text-editor__scroller">
                <div className="ids-rich-text-editor__editor" ref={onRef}>
                  <ContentEditable className="ids-rich-text-editor__input" />
                </div>
              </div>
            }
            placeholder={<Placeholder />}
            ErrorBoundary={LexicalErrorBoundary}
          />
          <HistoryPlugin />
          {autoFocus && <AutoFocusPlugin />}
          <DisablePlugin disabled={disabled} />
          <OnChangePlugin onChange={handleOnChange} ignoreSelectionChange />
          <OnFocusPlugin onFocus={handleOnFocus} onBlur={handleOnBlur} />
          <ListPlugin />
          <LinkPlugin />
          {floatingAnchorElem && (
            <FloatingLinkEditorPlugin
              anchorElem={floatingAnchorElem}
              messages={messages}
            />
          )}
          {isClearable && <ClearEditorPlugin />}
          <MarkdownShortcutPlugin transformers={TRANSFORMERS} />

          <CodeHighlightPlugin />
        </div>
        {showFooter && (
          <div className="ids-rich-text-editor__footer">
            {isPrivate && (
              <span className="ids-rich-text-editor__private">
                <Lock
                  size="small"
                  className="ids-rich-text-editor__private-icon"
                />
                {messages.private?.text}
              </span>
            )}
            {maxLength && (
              <>
                <CharacterLimitPlugin charset="UTF-8" maxLength={maxLength} />
                <MaxLengthPlugin maxLength={maxLength} />
              </>
            )}
            {primaryBtn &&
              React.cloneElement(primaryBtn, {
                className: cx(
                  'ids-rich-text-editor__primary-btn',
                  primaryBtn.props?.className
                ),
              })}
          </div>
        )}
      </div>
    </LexicalComposer>
  );
};

export default RichTextEditor;
