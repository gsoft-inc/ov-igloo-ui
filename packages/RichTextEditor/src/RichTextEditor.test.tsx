/**
  * @jest-environment jsdom
  */
 import React from 'react';
 import { render, screen } from '@testing-library/react';
 import MockTooltip from '@igloo-ui/tooltip/src/__mocks__/Tooltip.mock';

 import RichTextEditor from './RichTextEditor';

 jest.mock('@igloo-ui/tooltip', () => ({
  __esModule: true,
  default: jest.fn(MockTooltip),
}));

 const setup = (props = {}) => {
  return render(
    <RichTextEditor dataTest="ids-rich-text-editor" {...props} />
  );
};

 describe('RichTextEditor', () => {
   test('It should render without errors', () => {
    setup();
     const wrapper = screen.getByTestId('ids-rich-text-editor');
     expect(wrapper).toBeInTheDocument();
   });

   test('It should render a snapshot', () => {
     const {asFragment} = setup();
     expect(asFragment()).toMatchSnapshot();
   });
 });
