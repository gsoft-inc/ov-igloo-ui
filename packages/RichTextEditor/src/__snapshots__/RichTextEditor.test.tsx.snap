// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RichTextEditor It should render a snapshot 1`] = `
<DocumentFragment>
  <div
    class="ids-rich-text-editor"
    data-test="ids-rich-text-editor"
  >
    <div
      class="ids-toolbar"
    >
      <span
        class="ids-tooltip__container"
        role="button"
        tabindex="0"
      >
        <button
          class="ids-btn ids-icon-btn ids-icon-btn--small ids-toolbar__btn ids-btn--ghost-secondary"
          title=""
          type="button"
        >
          <span
            class=""
          >
            <svg
              class="ids-icon"
              fill="none"
              height="24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clip-rule="evenodd"
                d="M6 4.25a.75.75 0 0 0 0 1.5h1.25v12.5H6a.75.75 0 0 0 0 1.5h8a4.75 4.75 0 0 0 1.727-9.176A3.75 3.75 0 0 0 13 4.25H6Zm7 6a2.25 2.25 0 0 0 0-4.5H8.75v4.5H13Zm-4.25 1.5v6.5H14a3.25 3.25 0 0 0 0-6.5H8.75Z"
                fill="currentColor"
                fill-rule="evenodd"
              />
            </svg>
          </span>
        </button>
      </span>
      <span
        class="ids-tooltip__container"
        role="button"
        tabindex="0"
      >
        <button
          class="ids-btn ids-icon-btn ids-icon-btn--small ids-toolbar__btn ids-btn--ghost-secondary"
          title=""
          type="button"
        >
          <span
            class=""
          >
            <svg
              class="ids-icon"
              fill="none"
              height="24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clip-rule="evenodd"
                d="M17 4.25h-2.483a.76.76 0 0 0-.034 0H12a.75.75 0 0 0 0 1.5h1.436L8.97 18.25H7a.75.75 0 0 0 0 1.5h5a.75.75 0 0 0 0-1.5h-1.436l4.465-12.5H17a.75.75 0 0 0 0-1.5Z"
                fill="currentColor"
                fill-rule="evenodd"
              />
            </svg>
          </span>
        </button>
      </span>
      <span
        class="ids-tooltip__container"
        role="button"
        tabindex="0"
      >
        <button
          class="ids-btn ids-icon-btn ids-icon-btn--small ids-toolbar__btn ids-btn--ghost-secondary"
          title=""
          type="button"
        >
          <span
            class=""
          >
            <svg
              class="ids-icon"
              fill="none"
              height="24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clip-rule="evenodd"
                d="M7.25 12V5.75H6.5a.75.75 0 0 1 0-1.5h3a.75.75 0 0 1 0 1.5h-.75V12a3.25 3.25 0 0 0 6.5 0V5.75h-.75a.75.75 0 0 1 0-1.5h3a.75.75 0 0 1 0 1.5h-.75V12a4.75 4.75 0 1 1-9.5 0ZM5 18.25a.75.75 0 0 0 0 1.5h14a.75.75 0 0 0 0-1.5H5Z"
                fill="currentColor"
                fill-rule="evenodd"
              />
            </svg>
          </span>
        </button>
      </span>
      <span
        class="ids-tooltip__container"
        role="button"
        tabindex="0"
      >
        <button
          class="ids-btn ids-icon-btn ids-icon-btn--small ids-toolbar__btn ids-btn--ghost-secondary"
          title=""
          type="button"
        >
          <span
            class=""
          >
            <svg
              class="ids-icon"
              fill="none"
              height="24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clip-rule="evenodd"
                d="M12.117 5.75C10.14 5.75 8.75 7.09 8.75 8.5s1.389 2.75 3.367 2.75H19a.75.75 0 0 1 0 1.5h-3.4c.71.726 1.15 1.675 1.15 2.75 0 2.457-2.298 4.25-4.867 4.25-1.173 0-2.275-.366-3.133-.994V19a.75.75 0 0 1-1.5 0v-2.333a.75.75 0 0 1 1.443-.288c.43 1.037 1.651 1.871 3.19 1.871 1.978 0 3.367-1.34 3.367-2.75s-1.389-2.75-3.367-2.75H5a.75.75 0 0 1 0-1.5h3.4c-.71-.726-1.15-1.675-1.15-2.75 0-2.457 2.298-4.25 4.867-4.25 1.173 0 2.275.366 3.133.995V5a.75.75 0 0 1 1.5 0v2.333a.75.75 0 0 1-1.443.288c-.43-1.037-1.651-1.871-3.19-1.871Z"
                fill="currentColor"
                fill-rule="evenodd"
              />
            </svg>
          </span>
        </button>
      </span>
      <div
        class="ids-toolbar__divider"
      />
      <span
        class="ids-tooltip__container"
        role="button"
        tabindex="0"
      >
        <button
          class="ids-btn ids-icon-btn ids-icon-btn--small ids-toolbar__btn ids-btn--ghost-secondary"
          title=""
          type="button"
        >
          <span
            class=""
          >
            <svg
              class="ids-icon"
              fill="none"
              height="24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clip-rule="evenodd"
                d="M18.54 5.46a4.25 4.25 0 0 0-6.01 0l-1.06 1.06a.75.75 0 0 0 1.06 1.06l1.061-1.06a2.75 2.75 0 0 1 3.89 3.889l-2.24 2.24a2.583 2.583 0 0 1-3.653 0l-.118-.119a.75.75 0 0 0-1.061 1.061l.118.118a4.084 4.084 0 0 0 5.775 0l2.239-2.24a4.25 4.25 0 0 0 0-6.01ZM5.46 18.54a4.25 4.25 0 0 0 6.01 0l1.06-1.06a.75.75 0 0 0-1.06-1.06l-1.061 1.06a2.75 2.75 0 0 1-3.89-3.889l2.24-2.24a2.583 2.583 0 0 1 3.653 0l.118.119a.75.75 0 0 0 1.061-1.061l-.118-.118a4.083 4.083 0 0 0-5.775 0l-2.239 2.24a4.25 4.25 0 0 0 0 6.01Z"
                fill="currentColor"
                fill-rule="evenodd"
              />
            </svg>
          </span>
        </button>
      </span>
      <div
        class="ids-toolbar__divider"
      />
      <span
        class="ids-tooltip__container"
        role="button"
        tabindex="0"
      >
        <button
          class="ids-btn ids-icon-btn ids-icon-btn--small ids-toolbar__btn ids-btn--ghost-secondary"
          title=""
          type="button"
        >
          <span
            class=""
          >
            <svg
              class="ids-icon"
              fill="none"
              height="24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clip-rule="evenodd"
                d="M8.25 6A.75.75 0 0 1 9 5.25h10a.75.75 0 0 1 0 1.5H9A.75.75 0 0 1 8.25 6ZM5 4a.5.5 0 0 1 .5.5v3a.5.5 0 0 1-1 0v-3A.5.5 0 0 1 5 4ZM8.25 12a.75.75 0 0 1 .75-.75h10a.75.75 0 0 1 0 1.5H9a.75.75 0 0 1-.75-.75ZM4.606 11.167a.619.619 0 0 0-.106.333.5.5 0 0 1-1 0c0-.247.07-.606.294-.917C4.037 10.246 4.437 10 5 10s.963.246 1.206.583c.224.311.294.67.294.917 0 .414-.203.773-.422 1.047-.13.163-.282.315-.439.453H6a.5.5 0 0 1 0 1H4a.5.5 0 0 1-.224-.947l.003-.002.014-.007a4.487 4.487 0 0 0 .268-.148c.176-.103.409-.248.639-.421.233-.175.446-.364.597-.553.156-.195.203-.336.203-.422a.619.619 0 0 0-.106-.333C5.337 11.087 5.237 11 5 11c-.237 0-.337.088-.394.167ZM8.25 18a.75.75 0 0 1 .75-.75h10a.75.75 0 0 1 0 1.5H9a.75.75 0 0 1-.75-.75ZM5.915 17.292l.439-.438A.5.5 0 0 0 6 16H4a.5.5 0 0 0 0 1h.793l-.147.146a.5.5 0 0 0 .349.854H5v-.5l.354.354.561-.562ZM4 18.75h.5v-.014l.004.023a.267.267 0 0 0 .037.089c.021.031.105.152.459.152a.708.708 0 0 0 .295-.07.335.335 0 0 0 .135-.119.57.57 0 0 0 .07-.311.524.524 0 0 0-.072-.3.38.38 0 0 0-.152-.128.732.732 0 0 0-.281-.072H5m-1 .75h.5v-.016.002a.5.5 0 0 0-1 .014H4Zm1.915-1.458c.124.087.244.202.345.352.154.232.24.519.24.856 0 .33-.078.614-.224.845-.145.23-.338.38-.522.474C5.404 20 5.055 20 5 20H5c-.646 0-1.062-.254-1.291-.598a1.268 1.268 0 0 1-.208-.621v-.018l-.001-.007v-.005l.5-.001M5 17v.5-.5Z"
                fill="currentColor"
                fill-rule="evenodd"
              />
            </svg>
          </span>
        </button>
      </span>
      <span
        class="ids-tooltip__container"
        role="button"
        tabindex="0"
      >
        <button
          class="ids-btn ids-icon-btn ids-icon-btn--small ids-toolbar__btn ids-btn--ghost-secondary"
          title=""
          type="button"
        >
          <span
            class=""
          >
            <svg
              class="ids-icon"
              fill="none"
              height="24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clip-rule="evenodd"
                d="M5.25 6.75a1.25 1.25 0 1 0 0-2.5 1.25 1.25 0 0 0 0 2.5Zm3.75-2a.75.75 0 0 0 0 1.5h10a.75.75 0 0 0 0-1.5H9Zm0 6a.75.75 0 0 0 0 1.5h10a.75.75 0 0 0 0-1.5H9Zm-.75 6.75a.75.75 0 0 1 .75-.75h10a.75.75 0 0 1 0 1.5H9a.75.75 0 0 1-.75-.75Zm-1.75-6a1.25 1.25 0 1 1-2.5 0 1.25 1.25 0 0 1 2.5 0Zm-1.25 7.25a1.25 1.25 0 1 0 0-2.5 1.25 1.25 0 0 0 0 2.5Z"
                fill="currentColor"
                fill-rule="evenodd"
              />
            </svg>
          </span>
        </button>
      </span>
      <span
        class="ids-tooltip__container ids-toolbar__btn-clear-tooltip"
        role="button"
        tabindex="0"
      >
        <button
          class="ids-btn ids-toolbar__btn-clear ids-btn--ghost-secondary ids-btn--small"
          disabled=""
          title=""
          type="button"
        >
          <span
            class=""
          >
            Clear
          </span>
        </button>
      </span>
    </div>
    <div
      class="ids-rich-text-editor__inner"
    >
      <div
        class="ids-rich-text-editor__scroller"
      >
        <div
          class="ids-rich-text-editor__editor"
        >
          <div
            class="ids-rich-text-editor__input"
            contenteditable="true"
            data-lexical-editor="true"
            role="textbox"
            spellcheck="true"
            style="user-select: text; white-space: pre-wrap; word-break: break-word;"
          >
            <p
              class="ids-rich-text-editor__paragraph"
            >
              <br />
            </p>
          </div>
          <div
            class="ids-link-editor"
            style="opacity: 0; transform: translate(-10000px, -10000px);"
          />
        </div>
      </div>
      <div
        class="ids-rich-text-editor__placeholder"
      />
    </div>
  </div>
</DocumentFragment>
`;
